packages/gatsby-link/src/index.tsx

```diff
@@ -36,30 +48,39 @@ const createIntersectionObserver = (el, cb) => {
   return { instance: io, el }
 }
 
-function GatsbyLinkLocationWrapper(props) {
-  return (
-    <Location>
-      {({ location }) => <GatsbyLink {...props} _location={location} />}
-    </Location>
-  )
+interface IGatsbyLinkProps extends HTMLAttributes<HTMLElement> {
+  _location: {
+    pathname: string
+    search: string
+  }
+  innerRef: MutableRefObject<HTMLElement>
+  to: string
+  target: string
+  partiallyActive: boolean
+  activeClassName: string
+  activeStyle: CSSProperties
+  getProps: (arg0: unknown) => HTMLAttributes<HTMLElement> | null
+  replace: boolean
+  state: Record<string, unknown>
 }
 
-class GatsbyLink extends React.Component {
-  constructor(props) {
-    super(props)
-    // Default to no support for IntersectionObserver
-    let IOSupported = false
-    if (typeof window !== `undefined` && window.IntersectionObserver) {
-      IOSupported = true
-    }
+interface IGatsbyLinkState {
+  IOSupported: boolean
+}
 
-    this.state = {
-      IOSupported,
-    }
-    this.abortPrefetch = null
-    this.handleRef = this.handleRef.bind(this)
+class GatsbyLink extends React.Component<IGatsbyLinkProps, IGatsbyLinkState> {
+  readonly state: IGatsbyLinkState = {
```

Comment:
Two things here:

- Removing the constructor doesn't seem necessary, you can also type that one: https://fettblog.eu/typescript-react/components/#class-components
- According to https://react-typescript-cheatsheet.netlify.app/docs/basic/getting-started/class_components/ the `readonly` isn't necessary

So I'd go with the constructor for now to keep the functional changes minimal in this PR (more straight-forward conversion)